Add "show/hide captions" to <video> contextual menu when appropriate

danae404 opened the pull request:
#### c2541b657459f9c26fceb5ebf1d157b022e06ace
<pre>
Add &quot;show/hide captions&quot; to &lt;video&gt; contextual menu when appropriate
<a href="https://bugs.webkit.org/show_bug.cgi?id=260915">https://bugs.webkit.org/show_bug.cgi?id=260915</a>
rdar://14071575

Reviewed by NOBODY (OOPS!).

Adds a context menu item &quot;Show Subtitles (auto)&quot; / &quot;Hide Subtitles&quot; on
&lt;video&gt; elements on macOS that have a text track.

One new layout test.

* LayoutTests/http/tests/media/resources/hls/subtitles/subtitles.vtt: Added.
* LayoutTests/http/tests/media/track/toggle-show-subtitles-context-menu-item-expected.txt: Added.
* LayoutTests/http/tests/media/track/toggle-show-subtitles-context-menu-item.html: Added.
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::contextMenuItemSelected):
(WebCore::ContextMenuController::populate):
(WebCore::ContextMenuController::checkOrEnableIfNeeded const):
* Source/WebCore/platform/ContextMenuItem.cpp:
(WebCore::isValidContextMenuAction):
* Source/WebCore/platform/ContextMenuItem.h:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::contextMenuItemTagShowSubtitles):
(WebCore::contextMenuItemTagHideSubtitles):
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::videoHasSubtitles const):
(WebCore::HitTestResult::subtitlesAreShowing const):
(WebCore::HitTestResult::toggleShowSubtitlesForVideo const):
* Source/WebCore/rendering/HitTestResult.h:
* Source/WebKit/Shared/API/c/WKContextMenuItem.cpp:
* Source/WebKit/Shared/API/c/WKContextMenuItemTypes.h:
* Source/WebKit/Shared/API/c/WKSharedAPICast.h:
(WebKit::toAPI):
(WebKit::toImpl):
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiers.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiersPrivate.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c2541b657459f9c26fceb5ebf1d157b022e06ace

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17628 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17953 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18483 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19441 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16486 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17821 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21240 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18099 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18580 "12 flakes 180 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17839 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18138 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15303 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20288 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15373 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16048 "8 failures") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22643 "4 flakes 118 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16382 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16217 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20499 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16792 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14221 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15909 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/15889 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20275 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16635 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/en.lproj/Localizable.strings:
```diff
@@ -901,6 +901,12 @@
+"Show Subtitles" = "Show Subtitles";
+
+/* Hide Subtitles context menu item */
+"Hide Subtitles" = "Hide Subtitles";
```
Instead of inserting these by hand you can run `update-webkit-localizable-strings` (one benefit is that the script maintains the strings in sorted order).

danae404 replied:
Thanks!

aestes commented on Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiersPrivate.h:
```diff
@@ -68,5 +68,6 @@ WK_EXTERN NSString * const _WKMenuItemIdentifierCheckSpellingWhileTyping WK_API_
 WK_EXTERN NSString * const _WKMenuItemIdentifierCheckGrammarWithSpelling WK_API_AVAILABLE(macos(13.0), ios(16.0));
 WK_EXTERN NSString * const _WKMenuItemIdentifierPlayAllAnimations WK_API_AVAILABLE(macos(13.3), ios(16.4));
 WK_EXTERN NSString * const _WKMenuItemIdentifierPauseAllAnimations WK_API_AVAILABLE(macos(13.3), ios(16.4));
+WK_EXTERN NSString * const _WKMenuItemIdentifierToggleShowSubtitles WK_API_AVAILABLE(macos(14.0), ios(17.0));
```
This won't actually be available in macOS 14 / iOS 17, so you should use `WK_MAC_TBA` and `WK_IOS_TBA` here.

aestes commented on Source/WebCore/page/ContextMenuController.cpp:
```diff
@@ -374,6 +374,11 @@ void ContextMenuController::contextMenuItemSelected(ContextMenuAction action, co
     case ContextMenuItemTagToggleVideoEnhancedFullscreen:
         m_context.hitTestResult().toggleEnhancedFullscreenForVideo();
         break;
+#if PLATFORM(MAC)
```
Is the `#if PLATFORM(MAC)` needed here? The code it's guarding doesn't seem Mac-specific.

danae404 replied:
Fixed!

danae404 committed:
Add "show/hide captions" to <video> contextual menu when appropriate
https://bugs.webkit.org/show_bug.cgi?id=260915
rdar://14071575

Reviewed by NOBODY (OOPS!).

Adds a context menu item "Show Subtitles (auto)" / "Hide Subtitles" on
<video> elements on macOS that have a text track.

One new layout test.

* LayoutTests/http/tests/media/resources/hls/subtitles/subtitles.vtt: Added.
* LayoutTests/http/tests/media/track/toggle-show-subtitles-context-menu-item-expected.txt: Added.
* LayoutTests/http/tests/media/track/toggle-show-subtitles-context-menu-item.html: Added.
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::contextMenuItemSelected):
(WebCore::ContextMenuController::populate):
(WebCore::ContextMenuController::checkOrEnableIfNeeded const):
* Source/WebCore/platform/ContextMenuItem.cpp:
(WebCore::isValidContextMenuAction):
* Source/WebCore/platform/ContextMenuItem.h:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::contextMenuItemTagShowSubtitles):
(WebCore::contextMenuItemTagHideSubtitles):
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::videoHasSubtitles const):
(WebCore::HitTestResult::subtitlesAreShowing const):
(WebCore::HitTestResult::toggleShowSubtitlesForVideo const):
* Source/WebCore/rendering/HitTestResult.h:
* Source/WebKit/Shared/API/c/WKContextMenuItem.cpp:
* Source/WebKit/Shared/API/c/WKContextMenuItemTypes.h:
* Source/WebKit/Shared/API/c/WKSharedAPICast.h:
(WebKit::toAPI):
(WebKit::toImpl):
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiers.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiersPrivate.h:

eric-carlson commented on LayoutTests/http/tests/media/track/toggle-show-subtitles-context-menu-item.html:
```diff
@@ -0,0 +1,38 @@
+                x = video.offsetLeft + video.offsetWidth / 2;
+                y = video.offsetTop + video.offsetHeight / 2;
+                eventSender.mouseMoveTo(x,y);
+                eventSender.contextClick().find(item => item.title === "Show Subtitles (auto)").click();
```
It may be worth putting the result of `find(item => item.title === "Show Subtitles (auto)")` into variable so you can test for its presence separately, so the cause of the failure if it is somehow not in the menu will be more obvious.

danae404 replied:
Good idea!

eric-carlson commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -528,6 +529,48 @@ void HitTestResult::enterFullscreenForVideo() const
+bool HitTestResult::videoHasSubtitles() const
+{
+#if ENABLE(VIDEO)
+    if (HTMLMediaElement* mediaElement = this->mediaElement()) {
```
Nit: this can be `auto* mediaElement ...`

An <audio> element won't display subtitles, so I think this should also check to make sure this is a \<video> element.

eric-carlson commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -528,6 +529,48 @@ void HitTestResult::enterFullscreenForVideo() const
+#if ENABLE(VIDEO)
+    if (HTMLMediaElement* mediaElement = this->mediaElement()) {
+        if (auto* textTracks = mediaElement->textTracks())
+            return textTracks->length();
```
I think we only want to show the item if there is a caption or subtitles track with cues. This will show the item if an element has a metadata track, a chapter track, an empty captions track, etc.

eric-carlson commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -528,6 +529,48 @@ void HitTestResult::enterFullscreenForVideo() const
+bool HitTestResult::subtitlesAreShowing() const
+{
+#if ENABLE(VIDEO)
+    if (HTMLMediaElement* mediaElement = this->mediaElement()) {
```
Nit: auto

eric-carlson commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -528,6 +529,48 @@ void HitTestResult::enterFullscreenForVideo() const
+    if (HTMLMediaElement* mediaElement = this->mediaElement()) {
+        if (mediaElement->isVideo()) {
+            if (auto* textTracks = mediaElement->textTracks(); textTracks && textTracks->length()) {
+                for (unsigned i = 0; i < textTracks->length(); i++) {
+                    if (textTracks->item(i)->mode() == TextTrack::Mode::Showing)
+                        return true;
+                }
```
Ditto the comment above about only checking non-empty caption and subtitle tracks.

eric-carlson commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -528,6 +529,48 @@ void HitTestResult::enterFullscreenForVideo() const
+void HitTestResult::toggleShowSubtitlesForVideo() const
+{
+#if ENABLE(VIDEO)
+    if (HTMLMediaElement* mediaElement = this->mediaElement()) {
+        if (mediaElement->isVideo()) {
+            if (subtitlesAreShowing())
+                mediaElement->setSelectedTextTrack(&TextTrack::captionMenuOffItem());
+            else
+                mediaElement->setSelectedTextTrack(&TextTrack::captionMenuAutomaticItem());
+        }
+    }
```
I think this would be more idiomatic with early returns and a local variable:

```
    auto* mediaElement = this->mediaElement();
    if (!mediaElement || !mediaElement->isVideo())
        return;

    auto* selectedTrack = subtitlesAreShowing() ? ...
```
